element_zstack/bossdb.py

```diff
@@ -0,0 +1,161 @@
+        create_schema (bool): when True (default), create schema in the database if it
+                            does not yet exist.
+        create_tables (bool): when True (default), create schema tables in the database
+                             if they do not yet exist.
+        linking_module (str): A string containing the module name or module containing
+            the required dependencies to activate the schema.
+
+    Dependencies:
+    Tables:
+        volume.Volume: A parent table to VolumeUploadTask
+        volume.VoxelSize: A dependency of BossDBURLs
+    """
+
+    if isinstance(linking_module, str):
+        linking_module = importlib.import_module(linking_module)
+    assert inspect.ismodule(
+        linking_module
+    ), "The argument 'linking_module' must be a module's name or a module"
+
+    global _linking_module
+    _linking_module = linking_module
+
+    schema.activate(
+        schema_name,
+        create_schema=create_schema,
+        create_tables=create_tables,
+        add_objects=_linking_module.__dict__,
+    )
+
+
+@schema
+class VolumeUploadTask(dj.Manual):
+    """A pairing of Volume data to be uploaded and parameter set to be used for
+    the upload.
+
+    Attributes:
+        volume.Volume (foreign key): Primary key from `Volume`.
+        upload_type (str): One of 'image' (volumetric image) or 'annotation'
+        (segmentation data).
+        collection_name (varchar(64)): Name of the collection on bossdb.
+        experiment_name (varchar(64)): Name of the experiment on bossdb.
+        channel_name (varchar(64)): Name of the channel on bossdb.
+    """
+
+    definition = """
+    -> volume.Volume
+    upload_type='image': enum('image', 'annotation')
+    ---
+    collection_name: varchar(64)
+    experiment_name: varchar(64)
+    channel_name: varchar(64)
+    """
+
+
+@schema
+class BossDBURLs(dj.Imported):
+    """Uploads data to bossdb and stores the bossdb and neuroglancer URLs.
+
+    Attributes:
+        VolumeUploadTask (foreign key): Primary key from `VolumeUploadTask`.
+        bossdb_url (varchar(512)): bossdb URL for the uploaded data.
+        neuroglancer_url (varchar(1024)): neuroglancer URL for the uploaded data.
+    """
+
+    definition = """
+    -> VolumeUploadTask
+    ---
+    bossdb_url: varchar(512)
+    neuroglancer_url='': varchar(1024)
+    """
+
+    def get_neuroglancer_url(self, collection, experiment, channel):
+        base_url = f"boss://https://api.bossdb.io/{collection}/{experiment}/{channel}"
+        return (
+            "https://neuroglancer.bossdb.io/#!{'layers':{'"
+            + f"{experiment}"
+            + "':{'source':'"
+            + base_url
+            + "','name':'"
+            + f"{channel}"
+            + "'}}}"
+        )
+
+    @property
+    def key_source(self):
+        """Limit the upload to entries that have voxel sizes defined in the database."""
+        return VolumeUploadTask & volume.VoxelSize
```

Comment:
```suggestion
```

Reply:
How about we add `volume.VoxelSize` as a foreign key reference so the dependency on `volume.VoxelSize` is explicit?